### ISSUE TITLE
TypeError: data.copy is not a function

### ISSUE THREAD

**Issue by messa** [mongodb/js-bson#152]
Hello, I got this exception:

```
/lh-web/node_modules/mongodb/lib/mongo_client.js:447
              throw err
              ^
TypeError: data.copy is not a function
    at serializeBinary (/lh-web/node_modules/bson/lib/bson/parser/serializer.js:447:8)
    at serializeInto (/lh-web/node_modules/bson/lib/bson/parser/serializer.js:715:17)
    at [object Object].serialize (/lh-web/node_modules/bson/lib/bson/bson.js:47:27)
    at [object Object].Query.toBin (/lh-web/node_modules/mongodb-core/lib/connection/commands.js:123:25)
    at executeSingleOperation (/lh-web/node_modules/mongodb-core/lib/topologies/server.js:750:28)
    at [object Object].Server.command (/lh-web/node_modules/mongodb-core/lib/topologies/server.js:830:12)
    at executeScram (/lh-web/node_modules/mongodb-core/lib/auth/scram.js:175:14)
    at [object Object].ScramSHA1.auth (/lh-web/node_modules/mongodb-core/lib/auth/scram.js:275:5)
    at [object Object].Server.auth (/lh-web/node_modules/mongodb-core/lib/topologies/server.js:1043:40)
    at [object Object].Server.auth (/lh-web/node_modules/mongodb/lib/server.js:379:22)
    at authenticate (/lh-web/node_modules/mongodb/lib/db.js:1375:21)
    at [object Object].Db.authenticate (/lh-web/node_modules/mongodb/lib/db.js:1412:44)
    at /lh-web/node_modules/mongodb/lib/mongo_client.js:430:25
    at /lh-web/node_modules/mongodb/lib/db.js:229:5
    at [object Object].connectHandler (/lh-web/node_modules/mongodb/lib/server.js:279:7)
    at [object Object].g (events.js:260:16)
```

Please, do you have any idea where is the problem?

```shell
$ npm version
{ bson: '0.4.19',
  npm: '3.3.9',
  ares: '1.10.1-DEV',
  http_parser: '2.5.0',
  modules: '46',
  node: '4.2.1',
  openssl: '1.0.2d',
  uv: '1.7.5',
  v8: '4.5.103.35',
  zlib: '1.2.8' }
```

**Comment by christkv** [comment]
you are using the BSON binary type but not passing it a buffer object (probably a string)

**Comment by messa** [comment]
Got it. We are setting `window` in our node app:

```js
global.window = {};
```

If `window` is set, `node_modules/bson/lib/bson/binary.js` then works with `Uint8Array` instead of `Buffer` (because `Buffer` doesn't get defined):

```js
if(typeof window === 'undefined') {
  var Buffer = require('buffer').Buffer; // TODO just use global Buffer
}
// ...
function Binary(buffer, subType) {
    // ...
      if(typeof Buffer != 'undefined') {
        this.buffer = new Buffer(buffer);
      } else if(typeof Uint8Array != 'undefined' || (Object.prototype.toString.call(buffer) == '[object Array]')) {
        this.buffer = writeStringToArray(buffer);
```

The problem happens while authenticating against mongod, because in `node_modules/mongodb-core/lib/auth/scram.js` is the `Binary` created from a string:

```js
      // Build command structure
      var cmd = {
          saslStart: 1
        , mechanism: 'SCRAM-SHA-1'
        , payload: new Binary(f("n,,%s", firstBare))
        , autoAuthorize: 1
      }
```

where `var f = require('util').format`.

So - when string is passed to a Binary, it works if `window` is undefined - the string it is converted to `Buffer` (this is output of some debug logging I've put into `scram.js` to show the resulting `cmd`):

```
xxx ScramSHA1.auth cmd: { saslStart: 1,
  mechanism: 'SCRAM-SHA-1',
  payload: 
   Binary {
     _bsontype: 'Binary',
     sub_type: 0,
     position: 54,
     buffer: <Buffer 6e 2c 2c ... 54 32 2b ... > },
  autoAuthorize: 1 }
```

If `window` is defined:

```
xxx ScramSHA1.auth cmd: { saslStart: 1,
  mechanism: 'SCRAM-SHA-1',
  payload: 
   Binary {
     _bsontype: 'Binary',
     sub_type: 0,
     position: 54,
     buffer: 
      Uint8Array {
        '0': 110,
        '1': 44,
        '2': 44,
        '3': 110,
       ...
```

`Buffer` has a `copy` method, `Uint8Array` doesn't.

I think we can get rid of the global `window` so we can make it work. This problem hasn't occurred on development because we don't use MongoDB authorization while development and we apparently don't use Binary in our app. 

But I think that there is problem in bson implementation if, under some circumstances, `Binary` can fill itself with `Uint8Array` while it cannot be serialized.

You have written in the comment that we should pass buffer object, not string, into binary type, but it's the `mongodb-core` which is passing it a string.

**Comment by christkv** [comment]
I plan to introduce support for Uint8Array at some point soon but still need to figure out how to manage it without causing backwards breaking issues

**Comment by nicekiwi** [comment]
+1

**Comment by officert** [comment]
Is there a way to get around this? I am running into the same issue because I'm using the mongodb library in Electron where window is always defined (running node js inside Chrome).

**Comment by christkv** [comment]
you might need to use a Buffer implementation for the browser https://github.com/feross/buffer

**Comment by officert** [comment]
@christkv I'm not actually in a browser. I have the same problem as OP, running in Node (not browser) and the global window object is not null, which causes the ```Buffer`` module not to be required:

```
if(typeof window === 'undefined') {
  var Buffer = require('buffer').Buffer; // TODO just use global Buffer
}
```

If I simply change the file so it always requires ```Buffer``` everything works fine.

**Comment by christkv** [comment]
Not sure why window is defined, you must have some other module that is mocking up the global namespace with a fake window object.

**Comment by jonaswindey** [comment]
I'm having the same issue with debugging an application with iron-node.

Since iron-node uses Electron to run a node instance, Window is defined. This breaks connections to mongo 3.x when password authentication is used.

**Comment by kalahari** [comment]
@christkv In http://electron.atom.io/ a process is both node and a browser at the same time, so `window` is defined.

Is there any work-around that does not require patching the `js-bson` package as @officert has done?

**Comment by christkv** [comment]
@kalahari you are talking about the browser build ?, have you tried using just importing the bson parser itself ?

**Comment by kalahari** [comment]
I put together a quick demo project: https://github.com/kalahari/js-bson-issue152-electron-buffer
If you clone/download that and `npm install && npm start` it should open up a window and show you the error, somethign like this:

```
{ http_parser: '2.6.0',
  node: '5.1.1',
  v8: '4.7.80.27',
  uv: '1.7.5',
  zlib: '1.2.8',
  ares: '1.10.1-DEV',
  modules: '47',
  openssl: '1.0.2e',
  electron: '0.36.9',
  'atom-shell': '0.36.9',
  chrome: '47.0.2526.110' }
{ [MongoError: data.copy is not a function] name: 'MongoError', message: 'data.copy is not a function' }
TypeError: data.copy is not a function
    at serializeBinary (/private/tmp/js-bson-issue152-electron-buffer/node_modules/bson/lib/bson/parser/serializer.js:447:8)
    at serializeInto (/private/tmp/js-bson-issue152-electron-buffer/node_modules/bson/lib/bson/parser/serializer.js:715:17)
    at serialize (/private/tmp/js-bson-issue152-electron-buffer/node_modules/bson/lib/bson/bson.js:47:27)
    at Query.toBin (/private/tmp/js-bson-issue152-electron-buffer/node_modules/mongodb-core/lib/connection/commands.js:143:25)
    at executeSingleOperation (/private/tmp/js-bson-issue152-electron-buffer/node_modules/mongodb-core/lib/topologies/server.js:952:38)
    at Server.command (/private/tmp/js-bson-issue152-electron-buffer/node_modules/mongodb-core/lib/topologies/server.js:1038:3)
    at executeScram (/private/tmp/js-bson-issue152-electron-buffer/node_modules/mongodb-core/lib/auth/scram.js:176:14)
    at ScramSHA1.auth (/private/tmp/js-bson-issue152-electron-buffer/node_modules/mongodb-core/lib/auth/scram.js:276:5)
    at Server.auth (/private/tmp/js-bson-issue152-electron-buffer/node_modules/mongodb-core/lib/topologies/server.js:1167:40)
    at Server.auth (/private/tmp/js-bson-issue152-electron-buffer/node_modules/mongodb/lib/server.js:390:22)
```

All the action is in: https://github.com/kalahari/js-bson-issue152-electron-buffer/blob/master/window.js#L10 where the rejection happens at `db.authenticate()`.

```
"use strict";
var util = require("util");
var mongodb = require("mongodb");
function write(text) { window.document.getElementById("out").innerHTML += text + "\n"; }
window.document.addEventListener("DOMContentLoaded", function() {
    write(util.inspect(process.versions));
    var mc = new mongodb.MongoClient();
    mc.connect("mongodb://localhost")
        .then(function(db) {
            return db.authenticate("foo", "bar")
                .then(function() { return db.command({ connectionStatus : 1 }); })
                .then(function(status) { write(util.inspect(status)); });
        })
        .catch(function(err) { write(util.inspect(err) + "\n" + err.stack); });
});
```

This demo expects a mongod on localhost, auth doesn't need to be enabled, the error happens before credentials are sent.

**Comment by christkv** [comment]
fixed and in 0.4.22 in npm

**Comment by kalahari** [comment]
Confirmed the fix is working for me, much appreciated!
